### PR TITLE
Create test_crud_async.py

### PR DESCRIPTION
**Pull request by JericHunter**
Created an asynchronous version of tests for the CRUD test in the Cosmos SDK.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 1]
```diff
@@ -0,0 +1,2670 @@
...
+        offer_throughput = 1000
+        database_id = str(uuid.uuid4())
+        created_db = await self.client.create_database(
+            id=database_id,
+            offer_throughput=offer_throughput
+        )
+        self.assertEqual(created_db.id, database_id)
+
+        # Verify offer throughput for database
+        offer = created_db.read_offer()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:164]: missing await here

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 2]
```diff
@@ -0,0 +1,2670 @@
...
+        self.assertEqual(offer.offer_throughput, offer_throughput)
+
+        # Update database offer throughput
+        new_offer_throughput = 2000
+        offer = created_db.replace_throughput(new_offer_throughput)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:169]: miissing await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 3]
```diff
@@ -0,0 +1,2670 @@
...
+        try:
+            await func(*args, **kwargs)
+            self.assertFalse(True, 'function should fail.')
+        except exceptions.CosmosHttpResponseError as inst:
+            self.assertEqual(inst.status_code, status_code)
+
+    @classmethod
+    async def setUpClass(cls):
+        cls.client = cosmos_client.CosmosClient(cls.host, cls.masterKey, connection_policy=cls.connectionPolicy)
+        cls.database = await cls.client.create_database_if_nflot_exists(test_config._test_config.TEST_DATABASE_ID)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:103]: misspelled

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 4]
```diff
@@ -0,0 +1,2670 @@
...
+
+        # query with a string.
+        databases = [database async for database in
+                     self.client.query_databases('SELECT * FROM root r WHERE r.id="' + db2.id + '"')]  # nosec
+        self.assertEqual(1, len(databases), 'Unexpected number of query results.')
+        await self.client.delete_database(db1.id)
+        await self.client.delete_database(db2.id)
+
+    async def test_collection_crud(self):
+        created_db = self.databaseForTest
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:201]: we changed this name up top to just be database

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 5]
```diff
@@ -0,0 +1,2670 @@
...
+        collections = [collection async for collection in created_db.list_containers()]
+        # create a collection
+        before_create_collections_count = len(collections)
+        collection_id = 'test_collection_crud ' + str(uuid.uuid4())
+        collection_indexing_policy = {'indexingMode': 'consistent'}
+        created_recorder = RecordDiagnostics()
+        created_collection = created_db.create_container(id=collection_id,
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:208]: missing await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 6]
```diff
@@ -0,0 +1,2670 @@
...
+                                                         indexing_policy=collection_indexing_policy,
+                                                         partition_key=PartitionKey(path="/pk", kind="Hash"),
+                                                         response_hook=created_recorder)
+        self.assertEqual(collection_id, created_collection.id)
+        assert isinstance(created_recorder.headers, Mapping)
+        assert 'Content-Type' in created_recorder.headers
+        assert isinstance(created_recorder.body, Mapping)
+        assert 'id' in created_recorder.body
+
+        created_properties = created_collection.read()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:218]: missing await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 7]
```diff
@@ -0,0 +1,2670 @@
...
+
+            query='SELECT * FROM root r WHERE r.id=@id',
+            parameters=[
+                {'name': '@id', 'value': collection_id}
+            ]
+        )]
+
+        self.assertTrue(collections)
+        # delete collection
+        created_db.delete_container(created_collection.id)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:237]: missing await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 8]
```diff
@@ -0,0 +1,2670 @@
...
+        # read collection after deletion
+        created_container = created_db.get_container_client(created_collection.id)
+        await self.__AssertHTTPFailureWithStatus(StatusCodes.NOT_FOUND,
+                                                 created_container.read)
+
+        container_proxy = created_db.create_container_if_not_exists(id=created_collection.id,
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:243]: missing await

**Review by simorenoh — CHANGES_REQUESTED** [review]
Missing too many awaits still - please take your time to double check these things

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 9]
```diff
@@ -100,7 +100,7 @@ async def __AssertHTTPFailureWithStatus(self, status_code, func, *args, **kwargs
     @classmethod
     async def setUpClass(cls):
         cls.client = cosmos_client.CosmosClient(cls.host, cls.masterKey, connection_policy=cls.connectionPolicy)
-        cls.database = await cls.client.create_database_if_nflot_exists(test_config._test_config.TEST_DATABASE_ID)
+        cls.databaseForTest = cls.client.create_database_if_not_exists(test_config._test_config.TEST_DATABASE_ID)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:103]: still need the await here

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 10]
```diff
@@ -234,23 +234,23 @@ async def test_collection_crud(self):
 
         self.assertTrue(collections)
         # delete collection
-        created_db.delete_container(created_collection.id)
+        await created_db.delete_container(created_collection.id)
         # read collection after deletion
-        created_container = created_db.get_container_client(created_collection.id)
+        created_container = await created_db.get_container_client(created_collection.id)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:239]: does not need await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 11]
```diff
@@ -565,11 +565,11 @@ async def test_partitioned_collection_permissions(self):
                                'key': 1
                                }
 
-        all_collection.client_connection = restricted_client.client_connection
-        read_collection.client_connection = restricted_client.client_connection
+        all_collection.client_connection = await restricted_client.client_connection
+        read_collection.client_connection = await restricted_client.client_connection
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:569]: dont need awaits, this is just changing the reference from one to the other

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 12]
```diff
@@ -1188,7 +1188,7 @@ async def test_permission_crud(self):
         self.assertTrue(results)
 
         # replace permission
-        change_permission = permission.properties.copy()
+        change_permission = await permission.properties.copy()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1191]: doesn't need await - copy method

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 13]
```diff
@@ -1188,7 +1188,7 @@ async def test_permission_crud(self):
...
         permission.properties['id'] = 'replaced permission'
         permission.id = permission.properties['id']
         replaced_permission = user.replace_permission(change_permission['id'], permission.properties)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1194]: missing await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 14]
```diff
@@ -1357,17 +1357,17 @@ async def __SetupEntities(client):
...
             CRUDTests.host, resource_tokens, "Session", connection_policy=CRUDTests.connectionPolicy)
         db = entities['db']
 
-        old_client_connection = db.client_connection
-        db.client_connection = col_client.client_connection
+        old_client_connection = await db.client_connection
+        db.client_connection = await col_client.client_connection
         # 1. Success-- Use Col Permission to Read
-        success_coll = db.get_container_client(container=entities['coll'])
+        success_coll = await db.get_container_client(container=entities['coll'])
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1370]: doesn't need await - get_client() methods only make an object instance, not any network calls

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 15]
```diff
@@ -1404,14 +1404,15 @@ async def __SetupEntities(client):
...
+        read_doc = await doc_client.get_database_client(db.id).get_container_client(success_coll.id).read_item(docId,
+                                                                                                               docId)
         self.assertEqual(read_doc["id"], docId)
 
         # 6. Success-- Use Doc permission to delete doc
         await doc_client.get_database_client(db.id).get_container_client(success_coll.id).delete_item(docId, docId)
         self.assertEqual(read_doc["id"], docId)
 
-        db.client_connection = old_client_connection
+        db.client_connection = await old_client_connection
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1415]: no need for await, just changing a reference

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 16]
```diff
@@ -1455,9 +1456,9 @@ async def test_trigger_crud(self):
         self.assertTrue(triggers)
 
         # replace trigger
-        change_trigger = trigger.copy()
+        change_trigger = await trigger.copy()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1459]: no need for await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 17]
```diff
@@ -1512,19 +1513,19 @@ async def test_udf_crud(self):
         )]
         self.assertTrue(results)
         # replace udf
-        change_udf = udf.copy()
+        change_udf = await udf.copy()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1516]: no await needed

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 18]
```diff
@@ -1567,9 +1568,9 @@ async def test_sproc_crud(self):
         )]
         self.assertIsNotNone(sprocs)
         # replace sproc
-        change_sproc = sproc.copy()
+        change_sproc = await sproc.copy()
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1571]: no need for await

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 19]
```diff
@@ -1699,52 +1700,52 @@ async def test_collection_indexing_policy(self):
...
         # default policy
-        collection = db.create_container(
+        collection = await db.create_container(
             id='test_create_default_indexing_policy TestCreateDefaultPolicy03' + str(uuid.uuid4()),
             indexing_policy={},
             partition_key=PartitionKey(path='/id', kind='Hash')
         )
-        collection_properties = collection.read()
+        collection_properties = await collection.read()
         await self._check_default_indexing_policy_paths(collection_properties['indexingPolicy'])
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1744]: doesn't need await, and method doesn't need to be async

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 20]
```diff
@@ -1853,8 +1854,8 @@ async def test_create_indexing_policy_with_composite_and_spatial_indexes(self):
             logging_enable=True,
             logger=custom_logger,
         )
-        created_properties = created_container.read(logger=custom_logger)
-        read_indexing_policy = created_properties['indexingPolicy']
+        created_properties = await created_container.read(logger=custom_logger)
+        read_indexing_policy = await created_properties['indexingPolicy']
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1858]: doesn't need await, just grabbing a property from an existing object

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 21]
```diff
@@ -1877,7 +1878,7 @@ async def __get_first(array):
         # included paths should be 1: '/'.
         self.assertEqual(1, len(indexing_policy['includedPaths']))
 
-        root_included_path = __get_first([included_path for included_path in indexing_policy['includedPaths']
+        root_included_path = await __get_first([included_path for included_path in indexing_policy['includedPaths']
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1881]: __get_first shouldn't be async - look at what the method is doing

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 22]
```diff
@@ -0,0 +1,2671 @@
...
+    async def test_client_request_timeout(self):
+        # Test is flaky on Emulator
+        if not ('localhost' in self.host or '127.0.0.1' in self.host):
+            connection_policy = documents.ConnectionPolicy()
+            # making timeout 0 ms to make sure it will throw
+            connection_policy.RequestTimeout = 0.000000000001
+
+            with self.assertRaises(Exception):
+                # client does a getDatabaseAccount on initialization, which will time out
+                cosmos_client.CosmosClient(CRUDTests.host, CRUDTests.masterKey, "Session",
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1894]: async client needs to be initialized using async with here

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 23]
```diff
@@ -0,0 +1,2671 @@
...
+        connection_policy.ConnectionRetryConfiguration = Retry(
+            total=3,
+            read=3,
+            connect=3,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 504)
+        )
+        with self.assertRaises(AzureError):
+            # client does a getDatabaseAccount on initialization, which will time out
+            cosmos_client.CosmosClient(CRUDTests.host, CRUDTests.masterKey, "Session",
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1910]: same as above

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 24]
```diff
@@ -0,0 +1,2671 @@
...
+        retry_policy = Retry(
+            total=retries,
+            read=retries,
+            connect=retries,
+            backoff_factor=0.3,
+            status_forcelist=(500, 502, 504)
+        )
+        start_time = time.time()
+        try:
+            cosmos_client.CosmosClient(
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1932]: async with here too

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 25]
```diff
@@ -0,0 +1,2671 @@
...
+                connection_retry_policy=retry_policy)
+            self.fail()
+        except AzureError as e:
+            end_time = time.time()
+            return end_time - start_time
+
+    async def initialize_client_with_connection_core_retry_config(self, retries):
+        start_time = time.time()
+        try:
+            cosmos_client.CosmosClient(
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1945]: same as above

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 26]
```diff
@@ -0,0 +1,2671 @@
...
+                retry_connect=retries,
+                retry_status=retries)
+            self.fail()
+        except AzureError as e:
+            end_time = time.time()
+            return end_time - start_time
+
+    async def test_absolute_client_timeout(self):
+        with self.assertRaises(exceptions.CosmosClientTimeoutError):
+            cosmos_client.CosmosClient(
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:1960]: same here and line 1969

Comment (simorenoh) [reply sdk/cosmos/azure-cosmos/test/test_crud_async.py:1960]: and all client initializations below

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 27]
```diff
@@ -0,0 +1,2671 @@
...
+            """Creates resources for this test.
+
+            :Parameters:
+                - `client`: cosmos_client_connection.CosmosClientConnection
+
+            :Returns:
+                dict
+
+            """
+            collection = self.configs.create_multi_partition_collection_with_custom_pk_if_not_exist(self.client)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:2008]: this needs to be a new async method - any method in the self.configs is going to be for the sync client only
I would just do
```suggestion
            collection = await self.databaseForTest.create_container(<CONFIGS>)
```
where CONFIGS are the actual name and partition key being used in the original method in the configs file

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 28]
```diff
@@ -0,0 +1,2671 @@
...
+                'coll': collection,
+                'doc1': doc1,
+                'doc2': doc2,
+                'doc3': doc3
+            }
+            return resources
+
+        # Validate QueryIterable by converting it to a list.
+        resources = await __create_resources(self.client)
+        results = await resources['coll'].read_all_items(max_item_count=2)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:2022]: no need for await here - read_all_items is a query method

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 29]
```diff
@@ -0,0 +1,2671 @@
...
+        self.assertEqual(3,
+                         len(docs),
+                         'QueryIterable should return all documents' +
+                         ' using continuation')
+        self.assertEqual(resources['doc1']['id'], docs[0]['id'])
+        self.assertEqual(resources['doc2']['id'], docs[1]['id'])
+        self.assertEqual(resources['doc3']['id'], docs[2]['id'])
+
+        # Validate QueryIterable iterator with 'for'.
+        results = await resources['coll'].read_all_items(max_item_count=2)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:2033]: same here

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 30]
```diff
@@ -0,0 +1,2671 @@
...
+        counter = 0
+        # test QueryIterable with 'for'.
+        for doc in iter(results):
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:2036]: and this should be an async for loop over results

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 31]
```diff
@@ -0,0 +1,2671 @@
...
+                                 doc['id'],
+                                 'second document should be doc2')
+            elif counter == 3:
+                self.assertEqual(resources['doc3']['id'],
+                                 doc['id'],
+                                 'third document should be doc3')
+        self.assertEqual(counter, 3)
+
+        # Get query results page by page.
+        results = await resources['coll'].read_all_items(max_item_count=2)
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:2053]: query methods dont need await

**Review by simorenoh — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/test/test_crud_async.py [hunk 32]
```diff
@@ -0,0 +1,2647 @@
...
+
+        self.assertTrue(len(databases) > 0)
+
+        # read database.
+        self.client.get_database_client(created_db.id)
+
+        # delete database.
+        await self.client.delete_database(created_db.id)
+        # read database after deletion
+        read_db = self.client.get_database_client(created_db.id)
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/test/test_crud_async.py:133]: Q: I thought get_database_client will only get a databaseProxy, will it also read the database?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/test/test_crud_async.py:133]: gotcha, thanks

**Pull request merged**